### DISCUSSION TITLE
Second factor OTP work flow steps do not honor workflow order and other workflow issues.

### DISCUSSION THREAD

**Discussion by roytmana** [keycloak/keycloak#31782] [Q&A]
Hello Everyone 
I am new to keycloak and trying to prototype for our use case. We need email OTP as a second factor by default with optional SMS and Auth App for OTP. I create workflow where these 3 OTP options but even though email one is listed first it is always the SMS which is executed by default upon user/password login completion. If i remove DMS then it would be Auth App. No matter what I tried it is always that. Found some old github issue from 2022 about this and plan to implement concept of "preferred" step if not ordering but did not see any movement on it. Is there any way I can achieve what I need?

After failing with order of OTP steps , I tried to create two separate form workflows ("Authenticate using User/Password and email OTP vs Authenticate using User/Password and SMS OTP ) each starting with user/password and followed by its respective OTP step. That supported the order but when I clicked "Try Another Way" it presented both choices with the same name "Username and Password" not taking any name/description from my designed workflow. The workflow name seems to be coming from the first step (auth with username and password) which is hardcoded (looked from the code it seems to be pulling it from a resource bundle by key not from workflow design)
I also suspect it won't be easy to make it conditional based on whether user has configured SMS Phone or Auth App. 

I am at a loss as to what to do. This kind of things should be easy - maybe I am approaching it in a wrong way?

**Comment by darius-m** [comment]
Some additional information would be useful if people familiar with the email OTP or other plugins  can chime in:
 - what are the authentication plugins you are using, more specifially;
 - how the authentication flow is currently configured.

For the authentication flow you could use something like the following:
```
- username and password form (required)
- second authentication flow (required subflow)
  - email OTP (conditional)
  - ...
```

**Reply by roytmana** [reply]
Hello @darius-m thank you very much for reaching back to me!

I started with a very simple one. It disregards steps order and always execute SMS one while my requirement is default it to email
```
- username and password form (required)
- second authentication flow (required generic subflow)
  - Email TOTP Authentication (step alternative)
  - SMS Authentication (2FA) (step alternative)
  - OTP Form (step alternative)
```

then I tried to wrap each step into its own workflow. did not help - default was not email and order of alternative subflows was still not honored in the "try another way"
```
- username and password form (required)
- second authentication flow (required generic subflow)
   - Email workflow (alternative generic subflow)
      - Email TOTP Authentication (step alternative)
   - SMS workflow (alternative generic subflow)
      - SMS Authentication (2FA) (step alternative)
   - ORP Form workflow (alternative generic subflow)
      - OTP Form (step alternative)
```

I changed the above to have email, sms and otp form subflows conditional and added user configured to each - as expected it took me throug every one of then starting with email since they ar enot defined as alternatives and I do not know of any way to inform subsequeb steps of success of its predecessor. Of course "try another way" was not presented either

finally I tried two complete top level alternative workflows one password + email otp and the other password + sms
these worked by defaulting to password and email but if i click on "Try Another Way" it will **show two identical options** named "Username and PAssword" with no way to differentiate them

```
- password + email (alternative generic workflow)
  - username and password form (required)
  - second factor flow (conditional generic subflow)
       - User Configured (condition required)
       - Email TOTP Authentication (step required)
- password + sms (alternative generic workflow)
  - username and password form (required)
  - second factor flow (conditional generic subflow)
       - User Configured (condition required)
       - SMS Authentication (2FA) (step required)
```

I tried many other permutations but these above I repeated cleanly just for this post.

**Reply by darius-m** [reply]
I think the issue may be that the Email TOTP authentication step (which I assume is a third party plugin) does not have the correct functions implemented to validate whether it can be used when "Alternative" is set. There have been some discussions in the past about how alternatives should be used (and possibly enforced) - for example see #14988. The gist of it is that the subflows should be set to "Alternative", but the steps themselves should be "Required" - basically the second option that you have provided, but the subflow steps should be set to "Required".

When Keycloak evaluates the steps, it tries looking into the steps inside (I think the first one is the most important) to determine whether it can be executed. Since the Email OTP does not work correctly when outside of a subflow, it likely lacks the implementation to qualify it as an "alternative" step, so the entire subflow is deemed unusable; consequently, Keycloak just goes to the next one. In the last configuration that you have provided, the email TOTP is set to "Required", so it's probably why it works.

**Comment by roytmana** [comment]
Hello @darius-m ,

Sorry I had to switch to another project for some time. I reviewed code of both sms authenticator https://github.com/netzbegruenung/keycloak-mfa-plugins/blob/main/sms-authenticator/README.md and two email authenticators (https://github.com/mesutpiskin/keycloak-2fa-email-authenticator, https://github.com/5-stones/keycloak-email-otp) I tried comparing it with bundled App OTP one and could not find anything obviously lacking. Every configuration exhibits the same behavior - no ability to influence alternative authenticators order or at least identify a preferred one that should be executed by default. I tried to specify order property in authenticator factories, I looked at auth. flow's JSON in admin console for its property `priority`. The returned list of alt. authenticators (and thus default one) from a sub-flow disregards any changes in those two (factory order and flow priority)

Here are few relevant posts there were more of them struggling with alternative authenticators order/preferred 

Is there any prescribed way to achieve it? It seems to be such a common sense need and yet I could not find anything to address it properly 
There were several PRs floating around over the years. Has any one of them (either drop down for user to pick second factor during login or on profile or ordering of authenticators in workflow) went production and could be used?

https://github.com/keycloak/keycloak/issues/14340
https://github.com/keycloak/keycloak/pull/8397
https://github.com/keycloak/keycloak/issues/12102

**Reply by darius-m** [reply]
I'm not a developer, so I cannot say for sure, but I do remember some discussions mentioning that this may simply be an account UI issue, as reordering was allowed in one of the older account consoles (and there are admin REST API endpoints to increase / decrease the priority of an authenticator). You can almost certainly do that via the admin console, however. I don't know how it works, as I have not used this functionality (the first alternative is the one we expect as the "default" anyway).

**Comment by enggomarpu** [comment]
@roytmana 
Hope you are doing great. Did you able to solve your problem ?

- Username Password Form (Required)
- try-another-way forms (Required)
  - email subflow (Alternative)
    - Email OTP (Required)
  - sms subflow (Alternative)
    - SMS Authentication (Required)
    
    I am using this flow. 
    My problem is I want to show choiceScreen between Sms and Email after username and password. Do you know any way to do it ? I tried almost everything. I need 'try another way' functionality with choice screen.

**Reply by antonbakken** [reply]
@enggomarpu I am interested to see if you have found a solution to this issue? I am using pretty much the same flow as yours but would also like to have the 'try another way' functionality available.

**Comment by roytmana** [comment]
@enggomarpu
No we gave up on Keycloak for now
